viadot/sources/sap_rfc.py

```diff
@@ -463,37 +468,86 @@ def to_df(self):
         else:
             SEPARATORS = [sep]
 
-        records = None
         for sep in SEPARATORS:
             logger.info(f"Checking if separator '{sep}' works.")
-            df = pd.DataFrame()
+            df = pd.DataFrame(columns=columns)
             self._query["DELIMITER"] = sep
             chunk = 1
+            row_index = 0
             for fields in fields_lists:
                 logger.info(f"Downloading {chunk} data chunk...")
+                self._query["FIELDS"] = fields
                 try:
-                    self._query["FIELDS"] = fields
-                    try:
-                        response = self.call(func, **params)
-                    except ABAPApplicationError as e:
-                        if e.key == "DATA_BUFFER_EXCEEDED":
-                            raise DataBufferExceeded(
-                                "Character limit per row exceeded. Please select fewer columns."
-                            )
-                        else:
-                            raise e
-                    record_key = "WA"
-                    data_raw = response["DATA"]
-                    records = [row[record_key].split(sep) for row in data_raw]
-                    df[fields] = records
-                    chunk += 1
-                except ValueError:
-                    df = pd.DataFrame()
-                    continue
-        if not records:
-            logger.warning("Empty output was generated.")
-            columns = []
-        df.columns = columns
+                    response = self.call(func, **params)
+                except ABAPApplicationError as e:
+                    if e.key == "DATA_BUFFER_EXCEEDED":
+                        raise DataBufferExceeded(
+                            "Character limit per row exceeded. Please select fewer columns."
+                        )
+                    else:
+                        raise e
+                record_key = "WA"
+                data_raw = np.array(response["DATA"])
+
+                if row_index == 0:
+                    row_index = data_raw.shape[0]
+                    if row_index == 0:
+                        logger.warning(
+                            f"Empty output was generated for chunk {chunk} in columns {fields}."
+                        )
+                        chunk += 1
+                        continue
+                elif data_raw.shape[0] != row_index:
+                    data_raw = data_raw[:row_index]
+                    logger.warning(
+                        f"New rows were generated during the execution of the script. The table is truncated to the number of rows for the first chunk"
+                    )
```

Comment:
What is this new logic for? There should be a comment at the top explaining it since it's not self-evident.